jq255.md

```diff
@@ -0,0 +1,1425 @@
+into `s0` and `s1`, the double-and-add algorithm described in the
+previous section can be modified with *two* windows, the second window
+containing points `zeta(j*P)`, and only half the number of point
+doublings.
+
+## High-Level Protocols
+
+*High-level protocols* are signatures, key exchange and hash-to-group
+operations. We specify here how these operations are performed, and
+in particular all encoding rules and handling of invalid inputs.
+
+The primitives below use the [BLAKE2][] hash function, specifically
+its BLAKE2s variant with a 32-byte (256-bit) output. The pseudocode
+expects the following API (which is compatible with Python's standard
+`hashlib`):
+
+  - `sh = hashlib.blake2s()` creates an object `sh` that incarnates
+     a new BLAKE2s computation.
+
+  - The data to hash is provided in one or several chunks of arbitrary
+    length. Each chunk is a sequence of bytes.
+
+  - `sh.digest()` finalizes the computation and returns the 32-byte
+    BLAKE2s output.
+
+### Encoding Rules
+
+All encodings here are expressed in terms of *bytes*. We use the term
+"byte" to denote what is, formally speaking, an *octet* (a group of
+eight bits). A byte has a numerical value in the 0 to 255 range. The
+least significant bit of the byte has weight 1, while the most
+significant bit has weight 128.
+
+Encoding formats are fixed-size sequences of bytes. Within a sequence,
+the first byte has index 0, the next byte has index 1, and so on;
+thus, the last byte of a 32-byte sequence has index 31.
+
+#### Field Elements
+
+A field element `x` is encoded as a fixed 32-byte sequence using the
+unsigned little-endian convention:
+
+```python
+def field_encode(x):
+    return int(x).to_bytes(32, byteorder='little')
+```
+
+In other words:
+
+  - The field element is considered as an integer `z` in the `0` to `q-1`
+    range.
+  - For `j = 0` to `31`, output byte of index `j` is set to the
+    integer value `bb[j] = floor(z / (256^j)) mod 256`.
+
+Since `q` is slightly below `2^255`, the last output byte (index 31)
+necessarily has value at most 127, which means that the most significant
+bit of that byte is necessarily equal to 0.
+
+When decoding, the following rules MUST be observed:
+
+  - The decode MUST reject as invalid any input sequence whose length
+    is not exactly 32 bytes.
+  - The integer value `z` must be recomputed by taking into account
+    all input bytes completely. In particular, the most significant
+    bit of the last byte MUST NOT be ignored.
+  - The recomputed `z` MUST be checked to be lower than `q`. Any
+    implicit or explicit reduction modulo `q` is incorrect. If
+    `z >= q`, then the decoder MUST reject the input.
+
+The following function implements these rules:
+
+```python
+def field_decode(bb):
+    if len(bb) != 32:
+        return False
+    z = int.from_bytes(bb, byteorder='little')
+    if z >= q:
+        return False
+    return GFq(z)
```

Comment:
This is the first mention of this type/constructor

Reply:
`GFq()` is defined near the start of the "Group Element Operations" section (line 171 currently):

> `GFq(j)` converts the integer `j` to a field element by reducing `j` modulo `q`.

Reply:
Oh sorry, nevermind then